### PR TITLE
Update Mesa

### PR DESCRIPTION
**Pull request by valentindavid**
Here is an update for Mesa. Radeon RX Vega cards require this update.

### CODE REVIEW

**Comment by valentindavid** [comment]
Seems that Steam uses the wrong libstdc++ for LLVM.

On other systems it works, because the 32 bit library paths are registered through ldconfig and the Steam wrapper script builds a LD_LIBRARY_PATH with the system paths before fallback libraries.

I will try to make the link to LLVM from Mesa static. But the real fix would be to be able to move ldconfig cache into a writable part so pathes can be added.

**Comment by Barthalion** [comment]
Note that it will break runtime ABI as soname of all LLVM libraries changes.

**Comment by valentindavid** [comment]
Are there any application depending on LLVM apart from build time (i.e. compiling with clang)?

**Comment by Barthalion** [comment]
Maybe GNOME Builder? Hard to say what's out there outside Flathub.

**Comment by valentindavid** [comment]
Seems GNOME Builder may link to Clang. And since GNOME Builder is not in C++ but in C, it would be why they never encountered big problems with the static libstdc++.

I propose to have both version.

**Comment by alexlarsson** [comment]
Yeah, we cannot break ABI like this in a stable release, the llvm libraries is part of the ABI and they are definitely not stable between major releases.

I think the proper solution to this is to build mesa specific llvm versions that we link statically into mesa.

**Comment by alexlarsson** [comment]
Also, we explicitly link stdc++ into mesa due to requirements from steam (among others). Different libstc++ versions have different ABIs, and we can't have the GL driver (our mesa) break because LD_LIBRARY_PATH made it look up the libstdc++ from the steam runtime, which is different. That is a direct wish from valve. I guess we just need to make the mesa-specific llvm also link stdc++ statically.

**Comment by valentindavid** [comment]
The standard behavior of steam is as such:
 - Check for libraries in ld.so.cache and compare them with the Steam runtime. If a newer version is found in the runtime, then a symbolic link to that runtime library is added in a pinning path.
 - Take all the directories found for libraries in ld.so.cache. This constitutes the system paths.
 - Build LD_LIBRARY_PATH with first, the pinning path, then the system paths, then Steam runtime, then original LD_LIBRARY_PATH.

All other distributions have the 32 bit libstdc++ library path in the ld.so.conf. So for them it is not a problem, Steam will always pick the system library.

But for us, we cannot add the 32 bit libraries in ld.so.cache with flatpak, so Steam will never use the system libraries.

It is going to be a problem not just for libstdc++. glibc might have the problem at some point. The only safe way to do it now is to have everything static. I am not sure this is a good solution.

One a side note, I found some other problems with Compat32 when I tried to use "STEAM_RUNTIME=0". If you mount it in a different path that it was built, then libraries that require RPATH/RUNPATH will not be found. libpulse has a RUNPATH to /usr/lib/pulseaudio to find libpulsecommon-10.0.so. We should check for libraries and binaries in the platform images that have RPATH or RUNPATH and fix them.

**Comment by valentindavid** [comment]
I have a fix for Steam that depends on flatpak/flatpak#1048. When this is approved and merged there will be a pull request to flathub/com.valvesoftware.Steam.

During that time I will try to find a way to have both LLVM 4 and 5 in the runtime.

**Comment by valentindavid** [comment]
clang had "cleanup-platform": [ "*" ]

Should we expect some application copy /usr/lib/libclang* into /app/lib?

**Comment by alexlarsson** [comment]
Hmm, i worry a bit about the ABI compatibility of this. Given that we shipped llvm4 .so:s before, we have to keep those around, as existing binarlies may link to them. Also, 4 & 5 are not binary compatible, so anyone that links to both in the same process will crash and burn.

Issues I have with the above patch:

It puts llvm4 in a subdirectory, so existing things that link to llvm4 will not find it.
It links mesa dynamically against llvm5 so:s, so, any binary that also links to llvm4 will get two copies in the namespace and crash.
It still changes mesa to not use a static libstdc++, which can cause issues in the steam runtime.
Why do we need two clangs? clang is really only used as a devel tool, and for now i think keeping clang4 in the runtime is good enough.

So, instead of this patch I think what we want to do is build a completely static version of llvm5 that targets *only* the mesa targets (i.e. no arm/x86 backend), and link the new mesa against this.

**Comment by valentindavid** [comment]
Linking to libstdc++ in static crashes. We should not do that. For Steam I have solution involving proper handling of ld.so.conf. And I got this solution to full work with all games working like they should.
Do you know any distribution linking libstdc++ statically in Mesa?

**Comment by TingPing** [comment]
@valentindavid Fedora does.

**Comment by valentindavid** [comment]
Someone said maybe GNOME builder might be using clang as library.

I will gladely remove it.

Should I move llvm5 into a different prefix and keep llvm4 in /usr?

**Comment by alexlarsson** [comment]
fedora ships a statically linked libstdc++ mesa based on explicit requests from valve. I don't see why that should crash though.

gnome-builder does use libclang, but it uses the sdk as a runtime, so it will be there. Note: I'm not saying we shouldn't have clang in the sdk, just that we don't need both clang4 and clang5.

**Comment by alexlarsson** [comment]
Yeah, move the llvm5 .a files to a separate directory so nothing else picks them up by accident, and then remove them from the final build (so we don't have to support them).

**Comment by valentindavid** [comment]
T

**Comment by valentindavid** [comment]
It looks to me that Steam does not require this anymore since the default behavior now is to use the system libraries. This is since they have introduced STEAM_RUNTIME_PREFER_HOST_LIBRARIES.

**Comment by TingPing** [comment]
@valentindavid The flatpak doesn't use that, especially considering the 'host' doesn't have any of the libs.

**Comment by valentindavid** [comment]
I have checked SteamOS, they do not statically link. See http://repo.steampowered.com/steamos/pool/main/m/mesa/libgl1-mesa-dri_17.1.2-1+bsos1_amd64.deb file usr/lib/x86_64-linux-gnu/dri/radeonsi_dri.so links to libLLVM-4.0.so.1

Same with Debian.

**Comment by valentindavid** [comment]
From Mesa configure:
configure: error: DRI cannot be build as static library

**Comment by alexlarsson** [comment]
I don't mean that mesa itself should be a static library, that will not work. I mean that mesa should statically link to libllvm.a and libstdc++.a.

**Comment by alexlarsson** [comment]
I mean, the fact that our current mesa links dynamically to llvm is why all this is so hard. If the current mesa linked statically to llvm4, then we could just update it however we wanted without problems, because the exposed abi (opengl) is stable.

However, dynamically linking it exposes the entire llvm ABI (which is not stable) and makes it very hard to update it. In particular, there can never be another dynamic libllvm in the runtime, because if apps get both linked into the same process then calls to symbols in one will be redirected to incompatible ones in the other.

**Comment by alexlarsson** [comment]
@nwnk Hey, can you give a bit of feedback on this issue.

The issue is one of mesa vs static llvm and libstdc++ in the freedesktop runtime.
Unless i misremember things, valve recommended statically linking libstdc++ into mesa. Is this still true, and if so, how do we best do that? Additionally, we ship llvm 4 shared libs, and if we want to move our mesa to llvm5 we need to statically link that. Is that ok/safe?

**Comment by alexlarsson** [comment]
Hmm, we discussed on irc maybe not making the updated mesa part of the runtime, but instead a separate extension with a (basically) fully statically linked mesa (similar to nvidia deps) that would work on any version of the runtime. We could call it 1.4 like we do the nvidia one, and this could then continue to be updated over time and as it is updated for later runtime versions, it would also be used for older ones.

I guess that was a bit unclear, but essentially, we'd build a org.freedesktop.Platform.GL.default//1.4 extension. This will be automatically be picked up by everyone claiming to support the 1.4 GL extension point, which has been re-defined to mean "all future runtimes, but with no weird deps", to avoid having to have multiple versions of dependency-free things like nvidia.

**Comment by alexlarsson** [comment]
This could be used as a base:
https://github.com/flatpak/freedesktop-sdk-images/blob/1.6/org.freedesktop.Platform.GL.mesa-stable.json.in

**Comment by valentindavid** [comment]
Yes. At first I used that. But, it is not nice to use since you need to have an environment variable (which is undocumented by the way). Unless there is something detecting which drivers to use. But this detection will need to change when runtimes are changed. Owners of RX Vega cards need llvm 5.

**Comment by alexlarsson** [comment]
We don't need to set an environment variable if we call it ".default", because that is already in the `flatpak --gl-drivers` list by default. This was added for this specific usecase (updated gl drivers)

**Comment by valentindavid** [comment]
OK. I approve that then.

**Comment by valentindavid** [comment]
Tested as extension. The problem is the path to dri drivers.

Steam needs to set LIBGL_DRIVERS_PATH so that it includes the 32bit ones. It will set it to:
LIBGL_DRIVERS_PATH="/usr/lib/dri:/app/lib/32bit/lib/dri"

Unfortunately these are the wrong ones. Once they are set to the following:

LIBGL_DRIVERS_PATH=/usr/lib/GL/default/lib/dri:/app/lib/32bit/lib/GL/default/lib/dri

... It works.

The reason in the first place why LIBGL_DRIVERS_PATH needs to be set is because the 32 bit mesa has been installed into a different path than it is used. And it cannot be in the same path because it is conflicts with the path of the 64 bit mesa.

Maybe Compat32 and GL32 should not be a mere copy of the i386 build, but should be another build of it in a different path. There seem to be many problem with the copy.

**Comment by valentindavid** [comment]
This works if Steam defines: LIBGL_DRIVERS_PATH=/usr/lib/GL/default/lib/dri:/app/lib/32bit/lib/GL/default/lib/dri:/usr/lib/dri:/app/lib/32bit/lib/dri

**Comment by valentindavid** [comment]
These are the required changes to Steam:

```
diff --git a/com.valvesoftware.Steam.json b/com.valvesoftware.Steam.json
index 98be35f..c631b57 100644
--- a/com.valvesoftware.Steam.json
+++ b/com.valvesoftware.Steam.json
@@ -30,9 +30,7 @@
         "--extension=org.freedesktop.Platform.GL32=merge-dirs=vulkan/icd.d;glvnd/egl_vendor.d",
         "--extension=org.freedesktop.Platform.GL32=download-if=active-gl-driver",
         "--extension=org.freedesktop.Platform.GL32=enable-if=active-gl-driver",
-        "--env=LD_LIBRARY_PATH=/app/lib:/app/lib/32bit/lib",
-        "--env=LIBGL_DRIVERS_PATH=/usr/lib/dri:/app/lib/32bit/lib/dri",
-        "--env=STEAM_RUNTIME_PREFER_HOST_LIBRARIES=0"
+        "--env=LIBGL_DRIVERS_PATH=/usr/lib/GL/default/lib/dri:/app/lib/32bit/lib/GL/default/lib/dri:/usr/lib/dri:/app/lib/32bit/lib/dri"
     ],
     "build-options" : {
         "env": {
@@ -73,7 +71,10 @@
                 "mkdir -p /app/lib/32bit",
                 "ln -s /app/lib/32bit/lib/ld-linux.so.2 /app/lib/ld-linux.so.2",
                 "mkdir -p /app/share/appdata",
-                "cp com.valvesoftware.Steam.appdata.xml /app/share/appdata"
+                "cp com.valvesoftware.Steam.appdata.xml /app/share/appdata",
+                "mkdir -p /app/etc",
+                "echo /app/lib/32bit/lib >/app/etc/ld.so.conf",
+                "echo /app/lib/32bit/lib/pulseaudio >>/app/etc/ld.so.conf"
             ]
         }
     ]
```

It is not necessary to have flatpak/flatpak#1048 for steam.

**Comment by valentindavid** [comment]
I have tested llvmpipe. And it does not crash either with those changes. TingPing told me llvmpipe did not work.

**Comment by valentindavid** [comment]
* The case were llvmpipe did not work involved missing dri, which I did not test. Working on that but it is not related to this PR.

**Comment by valentindavid** [comment]
@Plagman, We would like your attention on this.

We are discussing here about whether statically linking libstdc++ in Mesa is still useful. It seems to me that this was not necessary since the introduction of the "STEAM_RUNTIME_PREFER_HOST_LIBRARIES". 

SteamOS does not seem to statically link libstdc++ in Mesa. Fedora still statically links libstdc++ into Mesa. This was done for Steam.

Could we know what is the current recommendation from Valve about linking libstdc++ statically into Mesa?

**Comment by alexlarsson** [comment]
I talked a bit to @nwnk about this on irc. And here is his (paraphrased) opinion:

> It doesn't make sense to link libstdc++ static really, the tools don't allow enough control for it to work correctly. Fedora stopped doing this in rawhide.
>
> For llvm the story is a bit more complicated. Upstream is planning to enable symbol versioning (essentially --default-symver=$llvm_version) starting from llvm6. This will make llvm6 and later version possible to link multiple times in the same process. However, that is not enough to make llvm6 to coexist with earlier versions that don't have symbol versions.
> 
> It might be possible to link llvm statically and libstdc++ dynamically, but you would perhaps need a version script or something to ensure that you're not re-exporting the llvm symbols.

I'm not sure the re-exporting is a problem in practice, since we're not rebuilding libGL/libOpenGL, so it depends on how the libglvnd dlopens the driver. From a quick look (but not verified) it seems to use RTLD_LOCAL, which I *believe* will cause it to not expose recursively pulled in symbols globally. The driver will however resolve previously linked in symbold (such as llvm ones) that are global, so unversioned llvm lookups in the driver may resolve to the llvm4 in the runtime rather than e.g. a bundled llvm5.

So, i believe we have two options here for a mesa GL extension. Either we link llvm statically into mesa, or we build our llvm5 with a --default-symver patch and pull it in via an rpath that nothing else picks up, and ensure that our mesa driver properly do all references to llvm via versioned symbols.

**Comment by valentindavid** [comment]
The current state of the PR is to have LLVM static, and libstdc++ shared. And I am fine with that. We can have llvm shared when LLVM 6 is released.

**Comment by valentindavid** [comment]
I have tested the last changes I have made. This PR is ready for review again.

File: mesa-fixup-EGL-Implement-the-libglvnd-interface-for-EGL-v.patch [hunk 1]
```diff
@@ -1,33 +0,0 @@
-From 2d4094c2caad6cef2f5544b3966fcc37ceb32036 Mon Sep 17 00:00:00 2001
```
Comment (alexlarsson) [inline mesa-fixup-EGL-Implement-the-libglvnd-interface-for-EGL-v.patch]: You can't delete this patch, its used by the main build :)

File: ld.so.conf [hunk 2]
```diff
@@ -0,0 +1,3 @@
+include /run/flatpak/ld.so.conf.d/*.conf
+include /app/etc/ld.so.conf
+/app/lib
```
Comment (alexlarsson) [inline ld.so.conf]: Do we need this? I mean, any flatpak that supports ld.so.cache will generate an identical one automatically.

File: mesa-EGL-Implement-the-libglvnd-interface-for-EGL-v2.patch [hunk 3]
```diff
@@ -1,3952 +0,0 @@
-From 97b447ee480a09a744094562da36b019a3eeb10f Mon Sep 17 00:00:00 2001
```
Comment (alexlarsson) [inline mesa-EGL-Implement-the-libglvnd-interface-for-EGL-v2.patch]: Same here, don't delete it.

File: org.freedesktop.Platform.GL.mesa-stable.json.in [hunk 4]
```diff
@@ -20,14 +19,71 @@
...
+            "build-options" : {
+                "cflags": "-g1",
+                "cxxflags": "-g1"
+            },
+            "cleanup-platform": ["*"],
+            "build-options" : {
+                "arch" : {
+                    "i386" : {
+                        "config-opts" : [
+                            "-DLLVM_TARGETS_TO_BUILD=X86;AMDGPU;NVPTX;BPF"
```
Comment (alexlarsson) [inline org.freedesktop.Platform.GL.mesa-stable.json.in]: This is not right, if this is only needed for mesa we shouldn't need X86 and BPF targets.
Same below.

File: org.freedesktop.Platform.GL.mesa-stable.json.in [hunk 5]
```diff
@@ -20,14 +19,71 @@
...
+                "-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON",
+                "-DCMAKE_BUILD_TYPE=RelWithDebInfo",
+                "-DLLVM_LIBDIR_SUFFIX=",
+                "-DLLVM_ENABLE_LIBCXX:BOOL=OFF",
+                "-DLLVM_ENABLE_ZLIB:BOOL=ON",
+                "-DLLVM_ENABLE_FFI:BOOL=ON",
+                "-DLLVM_ENABLE_RTTI:BOOL=ON",
+                "-DLLVM_BUILD_RUNTIME:BOOL=ON",
+                "-DLLVM_INCLUDE_TOOLS:BOOL=ON",
+                "-DLLVM_BUILD_TOOLS:BOOL=ON",
```
Comment (alexlarsson) [inline org.freedesktop.Platform.GL.mesa-stable.json.in]: Do we really need the tools?

File: ld.so.conf (hunk 2 above)
Comment (valentindavid) [reply ld.so.conf]: Yes. I think needed that for some testing.

Comment (valentindavid) [reply ld.so.conf]: We do not need it.

File: org.freedesktop.Platform.GL.mesa-stable.json.in (hunk 4 above)
Comment (valentindavid) [reply org.freedesktop.Platform.GL.mesa-stable.json.in]: I think X86 might be used by software rendering.

**Comment by alexlarsson** [comment]
This looks good to me to merge. I'll do some testing and eventually make the gnome sdk build it.
I'll squash all the changes though, there are just to many.

**Comment by alexlarsson** [comment]
Ugh, commited to the wrong branch, but i moved it manually right.

**Pull request merged**